custom_components/ocpp/ocppv16.py

```diff
@@ -0,0 +1,973 @@
+        else:
+            _LOGGER.warning("Failed with response: %s", resp.status)
+            await self.notify_ha(
+                f"Warning: Data transfer failed with response {resp.status}"
+            )
+            return False
+
+    async def get_configuration(self, key: str = ""):
+        """Get Configuration of charger for supported keys else return None."""
+        if key == "":
+            req = call.GetConfiguration()
+        else:
+            req = call.GetConfiguration(key=[key])
+        resp = await self.call(req)
+        if resp.configuration_key:
+            value = resp.configuration_key[0][om.value.value]
+            _LOGGER.debug("Get Configuration for %s: %s", key, value)
+            self._metrics[cdet.config_response.value].value = datetime.now(tz=UTC)
+            self._metrics[cdet.config_response.value].extra_attr = {key: value}
+            return value
+        if resp.unknown_key:
+            _LOGGER.warning("Get Configuration returned unknown key for: %s", key)
+            await self.notify_ha(f"Warning: charger reports {key} is unknown")
+            return None
+
+    async def configure(self, key: str, value: str):
+        """Configure charger by setting the key to target value.
+
+        First the configuration key is read using GetConfiguration. The key's
+        value is compared with the target value. If the key is already set to
+        the correct value nothing is done.
+
+        If the key has a different value a ChangeConfiguration request is issued.
+
+        """
+        req = call.GetConfiguration(key=[key])
+
+        resp = await self.call(req)
+
+        if resp.unknown_key is not None:
+            if key in resp.unknown_key:
+                _LOGGER.warning("%s is unknown (not supported)", key)
+                return
+
+        for key_value in resp.configuration_key:
+            # If the key already has the targeted value we don't need to set
+            # it.
+            if key_value[om.key.value] == key and key_value[om.value.value] == value:
+                return
+
+            if key_value.get(om.readonly.name, False):
+                _LOGGER.warning("%s is a read only setting", key)
+                await self.notify_ha(f"Warning: {key} is read-only")
+
+        req = call.ChangeConfiguration(key=key, value=value)
+
+        resp = await self.call(req)
+
+        if resp.status in [
+            ConfigurationStatus.rejected,
+            ConfigurationStatus.not_supported,
+        ]:
+            _LOGGER.warning("%s while setting %s to %s", resp.status, key, value)
+            await self.notify_ha(
+                f"Warning: charger reported {resp.status} while setting {key}={value}"
+            )
+
+        if resp.status == ConfigurationStatus.reboot_required:
+            self._requires_reboot = True
+            await self.notify_ha(f"A reboot is required to apply {key}={value}")
+
+    async def async_update_device_info_v16(self, boot_info: dict):
```

Comment:
It's a wrapper for async_update_device_info from base ChargePoint class. The reason this and async_update_device_info_v201 exist is because they each take BootNotification JSON object of respective OCPP version as an argument and those JSON objects have slightly different format.

Reply:
@lbbrhzn, what's your view? My logic is because these are not specific to the ocpp protocol they should be in the base class. The change in dict names between versions can be handled with an if statement or similar.